internal/pkg/manifest/union.go

```diff
@@ -62,48 +65,52 @@ func (t Union[_, _]) IsAdvanced() bool {
 }
 
 // UnmarshalYAML decodes value into either type Basic or Advanced, and stores that value
-// in t. value is first decoded into type Basic, and t will hold type Basic if:
-//   - There was no error decoding value into type Basic
-//   - Basic.IsZero() returns false OR Basic does not support IsZero()
+// in t. Value is first decoded into type Basic, and t will hold type Basic if
+// (1) There was no error decoding value into type Basic and
+// (2) Basic.IsZero() returns false OR Basic is not zero via reflection.
 //
 // If Basic didn't meet the above criteria, then value is decoded into type Advanced.
 // t will hold type Advanced if Advanced meets the same conditions that were required for type Basic.
 //
-// If value fails to decode into either type, or both types are zero after
-// decoding, t will not hold any value.
+// An error is returned if value fails to decode into either type
+// or both types are zero after decoding.
 func (t *Union[Basic, Advanced]) UnmarshalYAML(value *yaml.Node) error {
-	// reset struct
-	var simple Basic
-	var advanced Advanced
-	t.isBasic, t.Basic = false, simple
-	t.isAdvanced, t.Advanced = false, advanced
-
-	sErr := value.Decode(&simple)
-	if sErr == nil && !isZeroerAndZero(simple) {
-		t.isBasic, t.Basic = true, simple
+	var basic Basic
+	bErr := value.Decode(&basic)
+	if bErr == nil && !isZero(basic) {
+		t.SetBasic(basic)
 		return nil
 	}
 
+	var advanced Advanced
 	aErr := value.Decode(&advanced)
-	if aErr == nil && !isZeroerAndZero(advanced) {
-		t.isAdvanced, t.Advanced = true, advanced
+	if aErr == nil && !isZero(advanced) {
+		t.SetAdvanced(advanced)
 		return nil
 	}
 
-	if sErr != nil {
-		return sErr
+	// set an error to communicate why the Union is not
+	// of each type
+	if aErr == nil {
+		aErr = fmt.Errorf("is zero")
+	}
+	if bErr == nil {
+		bErr = fmt.Errorf("is zero")
 	}
-	return aErr
+
+	// multiline error because yaml.TypeError (which this likely is)
+	// is already a multiline error
+	return fmt.Errorf("%T: %s\n%T: %s", t.Basic, bErr, t.Advanced, aErr)
```

Comment:
I'm trying to think what are the scenarios that can happen here. Do you think this structure would be more helpful for customers?

```go
switch {
case aErr == nil && bErr == nil: // The basic form and advanced form were empty.
  return errors.New("ambiguous value: neither the simple or advanced form for the field was set")
case aErr != nil:
   return fmt.Errorf("unmarshal to %T advanced form: %w", t.Advanced, aErr)
default:
  return fmt.Errorf("unmarshal to %T basic form: %w", t.Basic, bErr)
}
```

Reply:
i like the proposed error first case (both values are zero), but the others i'm not quite as sold on 🤔. Say the union type is `Union[*string, aStruct]`. Given the yaml:
```
key:
  keyNotInaStruct: value
```
then `bErr == "unable to unmarshal !!map into string"` and `aErr == nil`. the proposed error would state `"unmarshal to string: <error>"`, which is true, but might be confusing along with the context that they were likely _trying_ to use the advanced form. i figured printing both the error for the string unmarshal _and_ the reasoning that the advanced type is zero might be more helpful - though it might be nice to include the "basic" and "advanced" type distinction. what do you think?

Reply:
oh I see what you mean 👍 that makes sense to me!!
Perhaps with just some slight tweak words but your reasoning sounds good:
```go
return fmt.Errorf("unmarshal to %T basic form: %v\nunmarshal to %T advanced form: %v", t.Basic, bErr, t.Advanced, aErr)
```